bottlecap/src/appsec/payload/mod.rs

```diff
@@ -0,0 +1,281 @@
+                    map[i] = ("server.request.path_params", path_params.to_waf_map()).into();
+                    i += 1;
+                }
+                if let Some(body) = body {
+                    map[i] = ("server.request.body", body).into();
+                    i += 1;
+                }
+                debug_assert_eq!(i, count); // Sanity check that we didn't over-allocate
+
+                map
+            }
+            HttpData::Response {
+                status_code,
+                headers,
+                body,
+                ..
+            } => {
+                let count = [status_code.is_some(), headers.is_some(), body.is_some()]
+                    .into_iter()
+                    .filter(|b| *b)
+                    .count();
+                let mut map = WafMap::new(count as u64);
+                let mut i = 0;
+
+                if let Some(response_status) = status_code {
+                    map[i] = ("server.response.status", response_status).into();
+                    i += 1;
+                }
+                if let Some(headers) = headers {
+                    let mut headers = headers.clone();
+                    headers.remove("set-cookie");
+                    map[i] = ("server.response.headers.no_cookies", headers.to_waf_map()).into();
+                    i += 1;
+                }
+                if let Some(response_body) = body {
+                    map[i] = ("server.response.body", response_body).into();
+                    i += 1;
+                }
+
+                debug_assert_eq!(i, count); // Sanity check that we didn't over-allocate
+
+                map
+            }
+        }
+    }
+}
+impl ToWafMap for HashMap<String, Vec<String>> {
+    fn to_waf_map(self) -> WafMap {
+        let mut map = WafMap::new(self.len() as u64);
+
+        for (i, (k, v)) in self.into_iter().enumerate() {
+            let mut arr = WafArray::new(v.len() as u64);
+            for (j, v) in v.into_iter().enumerate() {
+                arr[j] = v.as_str().into();
+            }
+
+            map[i] = (k.as_str(), arr).into();
+        }
+
+        map
+    }
+}
+impl ToWafMap for HashMap<String, String> {
+    fn to_waf_map(self) -> WafMap {
+        let mut map = WafMap::new(self.len() as u64);
+
+        for (i, (k, v)) in self.into_iter().enumerate() {
+            map[i] = (k.as_str(), v.as_str()).into();
+        }
+
+        map
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(super) enum RequestType {
+    APIGatewayV1, // Or Kong
+    APIGatewayV2Http,
+    APIGatewayV2Websocket,
+    APIGatewayLambdaAuthorizerToken,
+    APIGatewayLambdaAuthorizerRequest,
+    Alb,
+    LambdaFunctionUrl,
+}
+
+trait ExtractRequest {
+    const TYPE: RequestType;
+    async fn extract(self) -> HttpData;
+}
+trait ExtractResponse {
+    async fn extract(self) -> HttpData;
+}
+
+pub(super) async fn extract_request_address_data(body: &Bytes) -> Option<(HttpData, RequestType)> {
```

Comment:
I remember you mentioned maybe we want to eventually consolidate this with the triggers – overall, great

Deserializing into type might be slow since payloads can technically be up to 6MB

Reply:
I see that specific logic, as you mentioned privately, was ported from Go, maybe for the ones that already exist here in the triggers, could we use that same logic and for the events that we don't cover for inferred spans, I.E. Kong, we maintain the Go port?